cmd/controller/app/options/options.go

```diff
@@ -97,6 +98,8 @@ type ControllerOptions struct {
 	DefaultAutoCertificateAnnotations []string
 
 	// Allows specifying a list of custom nameservers to perform DNS checks on.
+	// For DNS lookups, this can be either specified like `<ip address>:<port>`.
+	// For DNS over HTTPS lookups, this can be specified like `https://<DoH RFC 8484 server address>`.
```

Comment:
I'd suggest rephrasing to make it clearer what the consequences having this flag has on cert-manager.

```suggestion
	// Allows specifying a list of custom nameservers to perform DNS checks on.
	// When this list is left empty, cert-manager works in self-recursion mode,
	// in which cert-manager tries to reproduce how ACME servers do the DNS-01
	// check. Conversely, when an item is given, cert-manager works in delegated
	// recursion mode, which means cert-manager lets the DNS server do the
	// recursion.
	//
	// Each nameserver can be either the IP address and port of a standard recursive
	// DNS server, or the endpoint to an RFC 8484 DNS over HTTPS endpoint. The
	// formats are:
	//  - Standard DNS:     8.8.8.8:53
	//  - DNS over HTTPS:   https://1.1.1.1/dns-query
```

Reply:
@maelvls
if the list is left empty, cert-manager will still try to fetch the authoritative results (this is the default value)

Reply:
if a value is provided, cert-manager will also still try to contact the authoritative nameservers; however, it will use the specified recursive DNS servers to determine what authoritative nameservers it should use.

Reply:
Only when `dns01-recursive-nameservers-only` is set to true, no authoritative nameservers are contacted.
I'm not sure about the recursion part, I think the authoritative server directly returns the correct result.

This is a bit confusing because we actually use the recursive DNS servers to find what authoritative DNS servers we should use, without recursing from the root.

Reply:
I think the above comment should be updated to reflect these three modes:

1. **Resolver mode**
   In this mode, cert-manager crawls up the tree using 8.8.8.8 but uses the first IP in the `NS` record to fetch the `TXT` record.
   - It starts by querying the SOA record for `_acme-challenge.domain.com.` using 8.8.8.8.
   - When it finds a `CNAME` it follows it.
   - When it finds an `SOA` record, it fetches the `NS` record using 8.8.8.8.
   - Finally, it uses one of the DNS IPs in there to fetch the `TXT` record.
2. **Resolver mode without authoritative check**:
   Same as "resolver mode" except cert-manager uses 8.8.8.8 for fetching the `TXT` record.
3. **Delegated mode** (`--dns01-recursive-nameservers-only=true`)
   In this mode, cert-manager directly queries the nameserver (8.8.8.8 by default).

In the three modes, it is possible to change the default 8.8.8.8 nameserver by using the flag `--dns01-recursive-nameservers`.

Is that how it works?

Reply:
@maelvls done. Looks good to me. Very precise. I've implemented the change. :)

Reply:
Sounds great, just for the record: the DoH DNS endpoint is also a recursive DNS server.

Reply:
@inteon Thanks for the info. I think it's good - he says "**standard** recursive DNS server" and "DoH Endpoint". Guess that's clear! :)

Reply:
Good point. `NS` servers (i.e., authoritative name servers) won't recursively resolve. 

@FlorianLiebhart My mistake, the comment isn't correct, it should be:

```suggestion
	// Allows specifying a list of custom nameservers to perform DNS checks on.
	//
	// Each nameserver can be either the IP address and port of a standard
	// DNS server or the endpoint to an RFC 8484 DNS over HTTPS endpoint.
	// For example, the following values are valid:
	//  - "8.8.8.8:53" (Standard DNS)
	//  - "https://1.1.1.1/dns-query" (DNS over HTTPS)
```